docs(models): Docathon (#202)
ecmwf/anemoi-core: models/docs/usage/create_model.rst

#########################
 Create your first model
#########################

This section describes how to create an existing model from the
``anemoi-models`` package.

In this example we show how to create an instance of the

> Encoder-Processor-Decoder that uses a Graphtransformer for the encoder

Reviewer: Minor: I think its "Graph Tranformer" not Graphtransformer

and decoder and a sliding window transformer [#f1]_ for the processor.

Our implemented models are instantiated by hydra [#f2]_. Commonly used

> model configurations can be found in ``configs/models`` (see generating

Reviewer: Can you add a link to "generating configs"? I guess you mean this here: [https://anemoi.readthedocs.io/projects/training/en/latest/start/hydra-intro.html#generating-user-config-files]

configs).

*****************************
 Loading Model Configuration
*****************************

> First, let's take a model configuration:

Reviewer: Maybe mention that this file is called ``transformer.yaml``

configuration for itself to understand the different bits needed to
create a model.

.. code:: python

   from omegaconf import OmegaConf

   model_config = OmegaConf.load("transformer.yaml")

********************************************************
 Getting statistics, data indices and supporting arrays
********************************************************

As described in :ref:`overview` we want to create a model interface that
can be used for training and inference. For that we need to create the
statistics, data indices and supporting arrays which is required for the
pre- and postprocessing.

The **statistics** are usually loaded from the dataset, i.e
``datamodule.statistics``. The statistics is simply a dictionary with
the mean, stdev, maximum and minimum of the variables.

.. code:: python

   statistics = {
       "mean": [0.5, 1.1, 0.0],
       "stdev": [0.1, 0.1, 0.1],
       "maximum": [1.0, 1.0, 1.0],
       "minimum": [0.0, 0.0, 0.0],
   }

The **data indices** are usually loaded from the dataset, i.e
``datamodule.data_indices``. The data indices is a dictionary with the
forcing and diagnostic variables.

.. code:: python

>    from anemoi.data.index import IndexCollection

Reviewer: If I follow just the installation from the anemoi-model documentation and I can't import with this `from anemoi.data.index import IndexCollection`. Did you mean `` from anemoi.models.data_indices.collection import IndexCollection``?

   name_to_index = {"10u": 0, "10v": 1, "2d": 2, "2t": 3}
   # This part is usually defined in the config/data/zarr.yaml file.
   data_config = dict(
       data={
           "forcing": ["cos_latitude"],
           "diagnostics": ["tp", "cp"],
           "remapper": [],
       }
   )
   data_indices = IndexCollection(data_config, name_to_index)

The **supporting arrays** are usually loaded from the dataset with
``datamodule.supporting_arrays``. The ``supporting_arrays`` is a
dictionary with the latitudes and longitudes of the grid.

.. code:: python

   supporting_arrays = {"latitudes": [90.0, 89.0, 88.0], "longitudes": [0.0, 1.0, 2.0]}

********************
 Creating the Graph
********************

All our currently implemented models are based on a graph encoder and
decoder. The graph is created by the ``GraphCreator`` class, see
:ref:`anemoi-graphs`.

.. code:: python

>    from anemoi.graphs.create import GraphCreator

Reviewer: Again, if I follow just the installation from the anemoi-model documentation and I can't import ``anemoi.graphs``

#########################
 Create your first model
#########################

This section describes how to create an existing model from the
``anemoi-models`` package.

In this example we show how to create an instance of the
Encoder-Processor-Decoder that uses a Graph Transformer for the encoder
and decoder and a sliding window transformer [#f1]_ for the processor.

Our implemented models are instantiated by hydra [#f2]_. Commonly used
model configurations can be found in ``configs/models`` (see `Generating

> user config files

Reviewer 2: I believe it's actually omegaconf and hydra

.. code:: yaml

   model:
     _target_: anemoi.models.models.encoder_processor_decoder.AnemoiModelEncProcDec

   layer_kernels:
       LayerNorm:
         _target_: torch.nn.LayerNorm
         _partial_: True
       Linear:
         _target_: torch.nn.Linear
         _partial_: True

   num_channels: 1024

   processor:
     _target_: anemoi.models.layers.processor.TransformerProcessor
     activation: GELU
     num_layers: 16
     num_chunks: 2

   encoder:
     _target_: anemoi.models.layers.mapper.GraphTransformerForwardMapper
     trainable_size: 8
     sub_graph_edge_attributes: ${model.attributes.edges}
     activation: GELU
     num_chunks: 1
     mlp_hidden_ratio: 4
     num_heads: 16

   decoder:
     _target_: anemoi.models.layers.mapper.GraphTransformerBackwardMapper
     trainable_size: 8
     sub_graph_edge_attributes: ${model.attributes.edges}
     activation: GELU
     num_chunks: 1
     mlp_hidden_ratio: 4
     num_heads: 16

   attributes:
     edges:
     - edge_length
     - edge_dirs
     nodes: []

Typically the model is instantiated in ``anemoi-training`` or
``anemoi-inference``. For this example we will load the model

> configuration for itself to understand the different bits needed to

Reviewer 2: ```suggestion
configuration by itself to understand the different components needed to
```

create a model.

.. code:: python

   from omegaconf import OmegaConf

   model_config = OmegaConf.load("transformer.yaml")

*******************************************************
 Define statistics, data indices and supporting arrays
*******************************************************

As described in :ref:`overview` we want to create a model interface that
can be used for training and inference. For that we need to create the
statistics, data indices and supporting arrays which is required for the
pre- and postprocessing.

The **statistics** are usually loaded from the dataset, i.e
``datamodule.statistics``. The statistics is simply a dictionary with
the mean, stdev, maximum and minimum of the variables.

.. code:: python

   statistics = {
       "mean": [0.5, 1.1, 0.0],
       "stdev": [0.1, 0.1, 0.1],
       "maximum": [1.0, 1.0, 1.0],
       "minimum": [0.0, 0.0, 0.0],
   }

The **data indices** are usually loaded from the dataset, i.e
``datamodule.data_indices``. The data indices is a dictionary with the

> forcing and diagnostic variables.

Reviewer 2: I think it would be good to reference and link to anemoi-datasets here.

Author: Added link to the anemoi-dataset docu.
